Fix vpn client

ersonp opened the pull request:
Did you run `make format && make check`?
yes

Fixes #783 
Fixes #900

 Changes:	
- Vpn retirer only trys 3 times
- Added errors from server hello to retire whitelist
- Removed conSummarry check in AppStates()
- Added new app state `AppStatusErrored` : ` "status": 2` which gives the error inside `detailed_status`
- Added a transport check before dailing routes adapted from [here](https://github.com/alexadhy/skywire/commit/916f90c73d79a6660cd17a16cb77b42bc645d843)
- Added `RPCErr` to preserve the type of the error we return via RPC (as suggested by @alexadhy) 

How to test this PR:
Test #783 
1. Run visorA with VPN server and passcode
2. Run visorB create a transport to vpn-server and connect to Vpn-server of visor A with the wrong passcode.
3. Visor won't retry after the first try.
4. Check `http://localhost:8000/api/visors/<pk>/summary` and see if `vpn-client` has
```json
{
  "name": "vpn-client",
  "args": [
    "-srv",
    "02d27b250feea1fc2fca518baff5eb30df0d77d43fd2cc323d1a25be6ce3abc7be"
  ],
  "auto_start": false,
  "port": 43,
  "status": 2,
  "detailed_status": "Password didn't match"
}
```

Test #900
1. Run visorB without creating a transport to vpn-server and connect to Vpn-server of visor A.
2. Visor won't retry after the first try.
3. Check `http://localhost:8000/api/visors/<pk>/summary` and see if `vpn-client` has
```json
{
  "name": "vpn-client",
  "args": [
    "-srv",
    "02d27b250feea1fc2fca518baff5eb30df0d77d43fd2cc323d1a25be6ce3abc7be"
  ],
  "auto_start": false,
  "port": 43,
  "status": 2,
  "detailed_status": "transport not found"
}
```

ersonp committed:
Update retrier

We update the retirer to not retry till success anymore. We only retry three times.

ersonp committed:
Add whitelisted errors

We whitelist errors sent from server hello as it's pointless to retry.

ersonp committed:
Remove conSummarry check

The commit contains changes to the logic of AppStates(). We are removing the connection summary check as whenever a vpn-client gets an error it closes all the connections to retry again. Because of this even when the vpn-client is still running AppStates() gives the app status for vpn-client as stopped. Insted depending on ProcByName() is sufficient.

ersonp committed:
Add app error

The commit contains additions to the proc and rpc code. The addition is of the app status AppStatusErrored that is set if there is an entry in statusErr which is set if the app encounters an error and stops. A . The commit only makes required changes in vpn-client to set the error.

ersonp committed:
Update MockRPCIngressClient

This commit contains updates to MockRPCIngressClient after the changes to RPCIngressClient in previous commit. The code was auto generated via mockery.

ersonp committed:
Add nolint:errcheck

This commit contains nolint:errcheck wherever the return value from a func was not checked to suppress warnings.

ersonp committed:
Remove test log

This commit removes log that was used for testing.

alexadhy commented:
CMIIW, I actually think that the `DetailedStatus` can be better with it being a struct containing the actual status of the running application (the enums like the current one in `develop`, as well as the current error description like the one you have). Can be beneficial also for tests.

So we can eliminate `SetDetailedStatusError` and instead just use the `SetDetailedStatus` for everything.

ersonp commented:
`SetDetailedStatusError` and `SetDetailedStatus` are separate because both of them are stored and fetched from different places, i.e. `detailed_status` is fetched from proc whereas the `error` is fetched from `procManager` as the proc is deleted when the app stops after an error. But I think it would be better to just have `SetDetailedStatus` just like you have suggested.

ersonp committed:
Change SetDetailedStatusError to SetError

The commit changes SetDetailedStatusError to SetError as it was confusing and SetError is nore consise and accurate.

ersonp commented:
@alexadhy I went through the code again and removing `SetDetailedStatusError`  would make the code needlessly complex so instead renamed all instances of `SetDetailedStatusError` to `SetError` which is more precise and easy to understand.

ersonp committed:
Fix func comment

ersonp committed:
Fix linting error

jdknives commented on internal/vpn/client.go:
```diff
@@ -188,16 +188,27 @@ func (c *Client) Serve() error {
-			fmt.Println("Connection broke, reconnecting...")
-			return fmt.Errorf("dialServeConn: %w", err)
+			switch err {
+			case errHandshakeStatusForbidden, errHandshakeStatusInternalError, errHandshakeNoFreeIPs,
```
Since the timeout is the only error we need to specifically check for redialing, would it be easier to check explicilty for `case errTimeout:`?

ersonp replied:
For that, we will have to make a `WithErrBlacklist` instead, as we are only exiting the retrier if the error is whitelisted.

jdknives commented on internal/vpn/client.go:
```diff
@@ -326,6 +337,11 @@ func (c *Client) setupTUN(tunIP, tunGateway net.IP) error {
 	tunIP, tunGateway, err := c.shakeHands(conn)
 	if err != nil {
+		switch err {
+		case errHandshakeStatusForbidden, errHandshakeStatusInternalError, errHandshakeNoFreeIPs,
```
Why dont we just return on all errors?

ersonp replied:
Did that so we could send `"error during client/server handshake: %w"` this extra error line if not for the above ones. Will take it out as it seems unnecessarry.

jdknives commented on internal/vpn/client.go:
```diff
@@ -445,6 +461,11 @@ func (c *Client) dialServeConn() error {
 	if err := c.serveConn(conn); err != nil {
+		switch err {
+		case errHandshakeStatusForbidden, errHandshakeStatusInternalError, errHandshakeNoFreeIPs,
+			errHandshakeStatusBadRequest:
```
Same question as above. Why are we checking for these errors here explicitly.

jdknives commented on internal/vpn/handshake_status.go:
```diff
@@ -32,3 +34,20 @@ func (hs HandshakeStatus) String() string {
+	case HandshakeStatusForbidden:
+		return errHandshakeStatusForbidden
+	default:
+		return errors.New("Unknown code")
```
Maybe `Unkown error code` is more explicit.

ersonp replied:
Okay.

jdknives commented on internal/vpn/handshake_status.go:
```diff
@@ -32,3 +34,20 @@ func (hs HandshakeStatus) String() string {
 	}
 }
+
+func (hs HandshakeStatus) getError() error {
```
Maybe I missed it but I could not find any reason we need the `func (hs HandshakeStatus) String() string {}`.

ersonp replied:
We were previously using that to print the handshake errors. And we are still using it to print server response
This is with the `String()` func
```bash
[proc:vpn-client:5d27178f0d224432a8bc92f10338e4f2]: Got server hello: {Forbidden <nil> <nil>}Closed TUN
```
And this is without the `String()` func
```bash
[proc:vpn-client:32f4cb6eb0684582810ef0979e95b514]: Got server hello: {4 <nil> <nil>}Closed TUN
```

jdknives commented on pkg/visor/api.go:
```diff
@@ -349,9 +350,22 @@ func (v *Visor) SetAppDetailedStatus(appName, status string) error {
 }
 
+// SetAppError implements API.
+func (v *Visor) SetAppError(appName, aErr string) error {
```
Maybe `appErr` is clearer since we call it `appName` alreadt?

ersonp replied:
Yes, sounds good.

jdknives reviewed (commented):
Just a few questions from my end. It seems like there may be a better way to handle processes so we do not need to set status and errors separately but its not the task of the PR to rewrite the whole library. So apart from the questions this looks good.

ersonp committed:
Add transport check

The commit contains a transport check in vpn-client. Before dialing the routes the available transports to the server are checked. If there is none then we send a ErrTransportNotFound. ErrTransportNotFound is whitelisted in the dail retrier.

ersonp committed:
Requested changes

The commit contains updated error for unknown code, aErr changed to appErr, removed switch statements and whitelisted ErrTransportNotFound in main retrier.

ersonp committed:
Make format

ersonp commented:
Should I add a max no of retries to the [retrier](https://github.com/ersonp/skywire/blob/abf6e3537c7539204a93681c568a1c9b8781fc15/internal/vpn/client.go#L119) of [`dialServer`](https://github.com/ersonp/skywire/blob/abf6e3537c7539204a93681c568a1c9b8781fc15/internal/vpn/client.go#L703) too?

ersonp committed:
Remove dialServer retrier

The commit removes the retrier in dialServer as it's redundant and the same job can be done by the main retrier.

ersonp committed:
Add RPCErr

The commit adds RPCErr struct that is used to preserve the type of the error from RPC. Currelty only the err of Dial() is wrapped with it.

ersonp committed:
Fix linting

mrpalide commented on internal/vpn/client.go:
```diff
@@ -188,16 +182,32 @@ func (c *Client) Serve() error {
+	r := netutil.NewRetrier(c.log, netutil.DefaultInitBackoff, netutil.DefaultMaxBackoff, 3, netutil.DefaultFactor).
+		WithErrWhitelist(errHandshakeStatusForbidden).WithErrWhitelist(errHandshakeStatusInternalError).
+		WithErrWhitelist(errHandshakeNoFreeIPs).WithErrWhitelist(errHandshakeStatusBadRequest).
+		WithErrWhitelist(errTransportNotFound)
```
`WithErrWhitelist` accept multiple error, so why didn't add all once and use multiple `WithErrWhitelist` method? Something like this:
```Go
r := netutil.NewRetrier(c.log, netutil.DefaultInitBackoff, netutil.DefaultMaxBackoff, 3, netutil.DefaultFactor).
	WithErrWhitelist(
		errHandshakeStatusForbidden,
		errHandshakeStatusInternalError,
		errHandshakeNoFreeIPs,
		errHandshakeStatusBadRequest,
		errTransportNotFound,
	)
```

ersonp replied:
That was an oversight on my part. Thanks for pointing that out.

mrpalide reviewed (commented):
Functionality work as well as possible, and code was clear, and just one little comment.

ersonp committed:
Requested changes

jdknives commented on pkg/router/router.go:
```diff
@@ -218,6 +218,11 @@ func (r *router) DialRoutes(
 	lPK := r.conf.PubKey
 	forwardDesc := routing.NewRouteDescriptor(lPK, rPK, lPort, rPort)
 
+	ok := r.checkIfTransportAvailabel(rPK)
```
This check should check if any transports are established for the visor and not if there is a transport between local and remote visor. Routes can be multihop and therefore this check in its current form is overly strict. Please change it and also fix the spelling issue with `Avalailable`.

jdknives replied:
The error consequently would be more appropriately named `ErrNoTransportFound`.

ersonp replied:
I feel like `ErrTransportNotFound` should be renamed to something like `ErrFailedToCreateRoute` which is more informative and `ErrNoTransportFound` and `ErrTransportNotFound` feel too similar.

ersonp committed:
Requested changes

The commit updates the checkIfTransportAvalailable to check if any transports are established for the visor and not if there is a transport between local and remote visor. checkIfTransportAvalailable also sends a new error ErrNoTransportFound if no transports are established for the visor.

Both ErrTransportNotFound and ErrNoTransportFound are whitelisted as ErrTransportNotFound is received from router whenever it is unsuccessful at creating a route to the remote visor..

ersonp committed:
Fix datarace

The commit changes the lock in SetError() from RLock to Lock.
